### ISSUE TITLE
ComboBox: Keyboard support needs minor improvements.

### ISSUE THREAD

**Issue by kdegrace** [IBM/carbon-components-react#1763]
There seems to be some minor issues with the ComboBox's keyboard support.

## Detailed description

When I first tab into the component, a focus outline is drawn around the control but the expected keyboard functions (e.g. space, up/down, etc) do not work.  The focus outline also has a strange little bump near the drop arrow. I suspect there's a missing `overflow:hidden` in there somewhere.

From here, if I tab a second time, the component still has focus and some of the keys work but I cannot type in the filter.  In this state, there is no focus outline so I can't visually tell if the control has focus.

From here, if I tab a third time, the input gets focus and I can type a filter.

> What did you expect to happen? What happened instead? What would you like to see changed?

I think the ideal scenario would be to go straight to that third tab state when the control gets focus.  In this state, I can type in the filter and also use the up/down arrows to select an item from the dropdown, etc.

> What browser are you working in?

Latest Chrome 

> What version of the Carbon Design System are you using?

v6.76.0

**Comment by carmacleod** [comment]
Please increase the priority of this one - these keyboard & accessibility problems aren't minor - they make the combobox really confusing to use unless you go all-mouse. Please see [this page that has 3 Carbon react comboboxes](https://cloud.ibm.com/status), and try using only the keyboard to interact with the 3 filters... now imagine using these with a screen reader and not being able to see the screen.

Please follow the pattern in [Example 3: List with Inline Autocomplete" of the ARIA Practices Guide](http://w3c.github.io/aria-practices/examples/combobox/aria1.1pattern/listbox-combo.html#ex3_label).

Some notes:
1. delete `tabindex="0"` from both of the outer divs, i.e. _only_ the input should take focus. So all keyboard events need to be on the input.
2. the `role="button"` needs to go on the div containing the triangle icon. Do not add a tabindex. Mouse events can go on this, but not keyboard events.
3. the `div.bx--list-box__field` should have `role="combobox"` and `aria-haspopup="listbox"`, `aria-expanded="true/false"`, and `aria-owns="unique-id-of-div.bx--list-box__menu"` (note: if using component-user-provided label for id, make sure id does not have spaces).
4. the `div.bx--form-item.bx--combo-box.bx--list-box` should not have any role or any ARIA related markup - it's just a generic container.
5. the `input` could have `role="textbox"` (which is the default for an input type="text", so technically, you can leave the role off), and it needs `aria-controls="unique-id-of-div.bx--list-box__menu"` (same id as for the aria-owns on the combobox div), `aria-autocomplete="list"` (because typing in the input filters the list), and `aria-label="Choose an item"` (delete `aria-label="ListBox input field"`). The input also needs `aria-activedescendant`, but your code there seems good.  :)
6. the `div.bx--list-box__menu` needs `role="listbox"` and the same aria-label as the input (i.e. `aria-label="Choose an item"`).

(just to note that the pattern in the APG shows a visible label, so if that is something that the Carbon component can do, then it needs to use `aria-labelledby="id-of-visible-label"` instead of `aria-label` on the input and the listbox).

**Comment by emyarod** [comment]
hi @kdegrace thanks for reporting, and thanks @carmacleod for the feedback! I am looking to incorporate these changes in #2113 for v10, and we most likely will be running additional audits to uncover usability issues with our components in the near future

**Comment by carmacleod** [comment]
> in the 3rd point, adding `role="combobox"` to the `<ListBox.Field>` element will throw a [warning](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/role-has-required-aria-props.md), seems like `aria-controls` is also expected?

The [spec](https://www.w3.org/TR/wai-aria-1.1/#combobox) for `role="combobox"` says:
> "When the combobox is expanded, authors MUST set aria-controls on the textbox element to a value that refers to the combobox popup element."

So, the warning is an eslint bug.

> in the 6th point, adding role="listbox" without a tabIndex would throw a [warning](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/interactive-supports-focus.md). would we also want to add a tabindex to div.bx--list-box__menu?

I'm not sure if it's necessary, because you are using aria-activedescendant on the input, but if it makes the warning go away, I guess it can't hurt to put tabindex="-1" on the listbox. You may also want style outline: none if you do that. (Ordinarily, I would never advise using outline: none without adding some other nice focus style, but the listbox really isn't supposed to take focus).

The ARIA 1.1 combobox pattern is new, and automated accessibility testing tools tend to get confused by it.  :)

I see that the PR for the vanilla combobox has been merged. Cool. I'll try it out tomorrow.  :)

**Comment by emyarod** [comment]
@carmacleod I've opened https://github.com/IBM/carbon-components-react/pull/2158 which should address the points you listed, would appreciate it if you could look that over! it won't contain the latest styles since there hasn't been a new vanilla Carbon release yet, but it will still have all of the markup changes so it would be great if you could review those changes
